src/Resources/contao/classes/ls_shop_product.php

```diff
@@ -2368,6 +2376,12 @@ public function getlinkToProduct($var_useVariantAliasOrID = '') {
             $addReturnPageToUrl = '/calledBy/searchResult';
         }
 
+        //redirect to productNotFound Published Page if no categorie/site is set for this product
```

Comment:
Ich habe das nicht durchdebuggt, es sieht für mich aber so aus, als würde in Zeilen 2370-2372 die aktuell aufgerufene Seite verwendet, sofern keine andere gefunden werden konnte. Es soll in diesem Fall ja aber die Default-Seite ("ls_shop_productNotFound" bzw. besser "ls_shop_defaultProductPage", siehe folgende Kommentare) verwendet werden. Das Produkt soll nur dann auf der aktuell aufgerufenen Seite dargestellt werden, wenn diese Seite dem Produkt auch zugeordnet ist.

Es sieht so aus, als ob `$objProductPage`, das in diesem Fall dann also die aktuell aufgerufene Seite ($objPage) ist, ignoriert wird, wenn in Zeile 2380 festgestellt wird, dass `$MainLanguagePageIDForLink` leer ist. Insofern könnte das, was ich im ersten Teil dieses Kommentars geschrieben habe, möglicherweise in der Praxis gar nicht dazu führen, dass ein Produkt auf der aktuellen Seite dargestellt wird, obwohl diese dem Produkt nicht zugeordnet ist. Es ist dann aber eigentlich unsauber, in Zeilen 2370 - 2372 einen Fallback anzuwenden, der in Zeilen 2380-2382 direkt wieder ausgehebelt wird.

Außerdem ist es meines Erachtens falsch, in der Zeile 2381 direkt die Default-Seite per return zurückzugeben. Man landet dann halt auf dieser Seite, aber es wird dann ja das Produkt dort überhaupt nicht dargestellt. Das Ziel ist es ja, die Darstellung des Produktes auf dieser Default-Seite zu ermöglichen. Deshalb ist es zwingend nötig, die URL auch im Falle der Default-Seite mit den Parametern für `product` und ggf. `selectVariant` zu erzeugen, wie es in Zeile 2385 geschieht.

Ich fände es des Weiteren richtig, auch im Falle der Default-Seite nicht einfach nur die URL per Return zurückzugeben (selbst wenn sie mit den genannten Parametern korrekt aufgebaut ist), sondern sie zuvor in `$this->ls_linkToProduct` zu schreiben und das dann zurückzugeben, so wie es in den Zeilen 2385 + 2387 geschieht. Wir müssen davon ausgehen, dass sich irgendwelche anderen Stellen im Merconis-Code auf dieses Verhalten verlassen und also erwarten, dass danach `$this->ls_linkToProduct` gesetzt und für den direkten Zugriff verfügbar ist.